blog/tanstack-router-intro/index.md

```diff
@@ -0,0 +1,461 @@
+
+```tsx
+export const Route = createFileRoute("/epics/$epicId/milestones/")({
+  component: ({}) => {
+    // ...
+```
+
+There's lots of other functions it supports. For search params we want `validateSearch`. This is our opportunity to tell Router _which_ search params this route supports, and how to validate what's currently in the url. After all, the user is free to type whatever they want into a url, regardless of the TypeScript typings you set up. It's your job to take potentially invalid values, and project them to something valid.
+
+First, let's define a type for our search params
+
+```ts
+type SearchParams = {
+  page: number;
+  search: string;
+  tags: string[];
+};
+```
+
+and now let's implement our `validateSearch` method. This receives a `Record<string, unknown>` representing whatever the user has in the url, and from that, we return something matching our type. Let's take a look.
+
+```tsx
+export const Route = createFileRoute("/epics/$epicId/milestones/")({
+  validateSearch(search: Record<string, unknown>): SearchParams {
+    return {
+      page: Number(search.page ?? "1") ?? 1,
+      search: (search.search as string) || "",
+      tags: Array.isArray(search.tags) ? search.tags : [],
+    };
+  },
+  component: ({}) => {
+```
+
+Note that (unlike `URLSearchParams`) we are not limited to just string values. We can put objects or arrays in there, and TanStack will do the work of serializing and de-serializing it for us. Not only that, but you can even specify [custom serialization mechanisms](https://tanstack.com/router/latest/docs/framework/react/guide/custom-search-param-serialization).
+
+Let's manually browse to this path, without any search params, and see what happens. When we browse to
+
+```
+http://localhost:5173/epics/1/milestones
+```
+
+Router replaces (does not redirect) us to
+
+```
+http://localhost:5173/epics/1/milestones?page=1&search=&tags=%5B%5D
+```
+
+TanStack ran our validation function, and then replaced our url with the correct, valid search params. If you don't like that it forces the url to be "ugly" like that, stay tuned; that's easily worked around. But first let's work with what we have.
+
+We've been using the `Route.useParams` method multiple times. There's also a `Route.useSearch` that does the same thing, for search params. But let's do something a little different. We've previously been putting everything in the same route file, so we could just directly reference the Route object from the same lexical scope. Let's build a separate component to read, and update these search params.
+
+I've added a `MilestoneSearch.tsx` component. You might think you could just import the `Route` object from the route file. But that's dangerous. You're likely to create a circular dependency, which might or might not work, depending on your bundler. And even if it "works" you might have some hidden issues lurking.
```

Comment:
w.r.t. "circular dependency":
I don't know if that is a problem at runtime, but it certainly is one at compile time, and TS compiler then rans into issues. That's why we advise against importing `Route` exports.